### DISCUSSION TITLE
Vim9: Class/Object member variable access control

### DISCUSSION THREAD

**Discussion by yegappan** [vim/vim#12979] [General]
Functions outside of a Vim9 class/object can have read-only access or a read-write access or no access
to a class/object member variable.   This is described in https://vimhelp.org/vim9class.txt.html.

The following sample code shows how to declare member variables with these different access controls:

```
vim9script
class A
    # object member: read-only access
    this.val1 = 10
    # object member: read/write access
    public this.val2 = 20
    # object member: private (no access)
    this._val3 = 30

    # class member: read-only access
    static val4 = 40
    # class member: read/write access
    public static val5 = 50
    # class member: private (no access)
    static _val6 = 60
endclass
```

This page is created to track the different ways these member variables can be accessed and whether they
are tested and verified to work properly.  The following cases are tested with the changes in the PR https://github.com/vim/vim/pull/12978.

**Access from script level:**

```
vim9script
class A
    this.ro_obj_var = 10
    public this.rw_obj_var = 20
    this._priv_obj_var = 30

    static ro_class_var = 40
    public static rw_class_var = 50
    static _priv_class_var = 60
endclass

var a = A.new()
echo a.ro_obj_var		# Works
a.ro_obj_var = 0		# Fails with E1335 (expected)
echo a.rw_obj_var		# Works
a.rw_obj_var = 0		# Works
echo a._priv_obj_var		# Fails with E1333 (expected)
a._priv_obj_var = 0		# Fails with E1333 (expected)

echo A.ro_class_var		# Works
A.ro_class_var = 0		# Fails with E1335 (expected)
echo A.rw_class_var		# Works
A.rw_class_var = 0		# Works
echo A._priv_class_var		# Fails with E1333 (expected)
A._priv_class_var = 0		# Fails with E1333 (expected)
```

**Access from a def function:**
```
vim9script
class A
    this.ro_obj_var = 10
    public this.rw_obj_var = 20
    this._priv_obj_var = 30

    static ro_class_var = 40
    public static rw_class_var = 50
    static _priv_class_var = 60
endclass

def T()
    var a = A.new()
    echo a.ro_obj_var		# Works
    a.ro_obj_var = 0		# Fails with E1335 (expected)
    echo a.rw_obj_var		# Works
    a.rw_obj_var = 0		# Works
    echo a._priv_obj_var	# Fails with E1333 (expected)
    a._priv_obj_var = 0		# Fails with E1333 (expected)
    
    echo A.ro_class_var		# Works
    A.ro_class_var = 0		# Fails with E1089 (BUG)
    echo A.rw_class_var		# Works
    A.rw_class_var = 0		# Fails with E1089 (BUG)
    echo A._priv_class_var	# Works (BUG)
    A._priv_class_var = 0	# Fails with E1089 (BUG)
enddef
T()
```

**Access from a method in the class:**
```
vim9script
class A
    this.ro_obj_var = 10
    public this.rw_obj_var = 20
    this._priv_obj_var = 30

    static ro_class_var = 40
    public static rw_class_var = 50
    static _priv_class_var = 60

    def Foo()
	echo this.ro_obj_var	# Works
	this.ro_obj_var = 0	# Works
	echo this.rw_obj_var	# Works
	this.rw_obj_var = 0	# Works
	echo this._priv_obj_var	# Works
	this._priv_obj_var = 0	# Works
	
	echo ro_class_var	# Works
	ro_class_var = 0	# Works
	echo rw_class_var	# Works
	rw_class_var = 0	# Works
	echo _priv_class_var	# Works
	_priv_class_var = 0	# Works

	echo A.ro_class_var	# Works
	A.ro_class_var = 0	# Fails with E1089 (BUG)
	echo A.rw_class_var	# Works
	A.rw_class_var = 0	# Fails with E1089 (BUG)
	echo A._priv_class_var	# Works
	A._priv_class_var = 0	# Fails with E1089 (BUG)
    enddef
endclass

var a = A.new()
a.Foo()
```

**Access from a method in a different class:**
```
vim9script
class A
    this.ro_obj_var = 10
    public this.rw_obj_var = 20
    this._priv_obj_var = 30

    static ro_class_var = 40
    public static rw_class_var = 50
    static _priv_class_var = 60
endclass

class B
    def Foo()
	var a = A.new()
	echo a.ro_obj_var	# Works
	a.ro_obj_var = 0	# Fails with E46 (expected)
	echo a.rw_obj_var	# Works
	a.rw_obj_var = 0	# Works
	echo a._priv_obj_var	# Fails with E1333 (expected)
	a._priv_obj_var = 0	# Fails with E1333 (expected)
	
	echo A.ro_class_var	# Works
	A.ro_class_var = 0	# Fails with E1089 (BUG)
	echo A.rw_class_var	# Works
	A.rw_class_var = 0	# Fails with E1089 (BUG)
	echo A._priv_class_var	# Works (BUG)
	A._priv_class_var = 0	# Fails with E1089 (BUG)
    enddef
endclass

var b = B.new()
b.Foo()
```

**Access from child class extending a class:**
```
vim9script
class A
    this.ro_obj_var = 10
    public this.rw_obj_var = 20
    this._priv_obj_var = 30

    static ro_class_var = 40
    public static rw_class_var = 50
    static _priv_class_var = 60
endclass

class B extends A
    def Foo()
	echo this.ro_obj_var	# Works
	this.ro_obj_var = 0	# Works
	echo this.rw_obj_var	# Works
	this.rw_obj_var = 0	# Works
	echo this._priv_obj_var	# Works
	this._priv_obj_var = 0	# Works
	
	echo ro_class_var	# Works
	ro_class_var = 0	# Works
	echo rw_class_var	# Works
	rw_class_var = 0	# Works
	echo _priv_class_var	# Works
	_priv_class_var = 0	# Works

	echo A.ro_class_var	# Works
	A.ro_class_var = 0	# Fails with E1089 (BUG)
	echo A.rw_class_var	# Works
	A.rw_class_var = 0	# Fails with E1089 (BUG)
	echo A._priv_class_var	# Works
	A._priv_class_var = 0	# Fails with E1089 (BUG)
    enddef
endclass

var b = B.new()
b.Foo()
```

**Access from a class implementing an interface:**
```
vim9script
interface A
    this.ro_obj_var: number
    public this.rw_obj_var: number
    this._priv_obj_var: number

    static ro_class_var: number
    public static rw_class_var: number
    static _priv_class_var: number
endinterface

class B implements A
    this.ro_obj_var = 10
    public this.rw_obj_var = 20
    this._priv_obj_var = 30

    static ro_class_var = 40
    public static rw_class_var = 50
    static _priv_class_var = 60

    def Foo()
	echo this.ro_obj_var	# Works
	this.ro_obj_var = 0	# Works
	echo this.rw_obj_var	# Works
	this.rw_obj_var = 0	# Works
	echo this._priv_obj_var	# Works
	this._priv_obj_var = 0	# Works
	
	echo ro_class_var	# Works
	ro_class_var = 0	# Works
	echo rw_class_var	# Works
	rw_class_var = 0	# Works
	echo _priv_class_var	# Works
	_priv_class_var = 0	# Works

	echo A.ro_class_var	# Works
	A.ro_class_var = 0	# Fails with E1089 (BUG)
	echo A.rw_class_var	# Works
	A.rw_class_var = 0	# Fails with E1089 (BUG)
	echo A._priv_class_var	# Works
	A._priv_class_var = 0	# Fails with E1089 (BUG)
    enddef
endclass

var b = B.new()
b.Foo()
```

_Note: The last test crashes Vim._

**Comment by stefanos82** [comment]
If I may ask, what was the reasoning behind the use of underscore for private member variables and not the use of `private` keyword?

Was it a matter of aesthetics or inspired by Python pseudo-privacy mechanism?

**Reply by yegappan** [reply]
The rationale for using underscore in private class/object member variable names is described in the section titled "Making object members private with an underscore" in https://vimhelp.org/vim9class.txt.html#_rationale

**Comment by stefanos82** [comment]
For some reason the use of underscore from an aesthetic point of view does not feel "right", but if it works... :unamused: 

Something like the following examples make it more obvious to my tired eyes that they mean `private member`:

```log

# Version 1

vim9script
class A
    this::val = 10 # "::" for private members
endclass

var a = A.new()
echo a::val	# Fails with E1333 (expected)
a::val = 20	# Fails with E1333 (expected)

# Version 2

vim9script
class A
    this.$val = 10 # "$" for private members
endclass

var a = A.new()
echo a.$val	# Fails with E1333 (expected)
a.$val = 20	# Fails with E1333 (expected)
```

Maybe it's not clear with the aforementioned examples, but with the use of private methods inside classes, it makes it clearer:

```log

# Original before from https://vimhelp.org/vim9class.txt.html#_rationale

class SomeClass
	def _Foo(): number
	  return 10
	enddef
	def Bar(): number
	  return this._Foo()
	enddef
endclass

# Version 1
class SomeClass
	def this::Foo(): number
	  return 10
	enddef
	def Bar(): number
	  return this::Foo()
	enddef
endclass

# Version 2
class SomeClass
	def $Foo(): number
	  return 10
	enddef
	def Bar(): number
	  return this.$Foo()
	enddef
endclass
```

**Reply by chrisbra** [reply]
whether we use `_` or `::` or `$` for private members seems like bike-shedding. And that decision has already been made 🤷

**Reply by stefanos82** [reply]
Oh well... :shrug:

**Comment by errael** [comment]
@yegappan Can you suggest something or related somethings that you'd expect to take less than a day to fix?

**Reply by yegappan** [reply]
Thanks.  The test for using the member variables from a class implementing an interface is crashing Vim (you need to comment out the lines that generate error).  Can you look into this? 

Another one is a private class member variable is accessible from a method in a different class.

**Reply by errael** [reply]
This simpler test fails
```
vim9script
interface A
    static ro_class_var: number
endinterface

class B implements A
    static ro_class_var = 40

    def Foo()
	var xxx = A.ro_class_var	# Works
    enddef
endclass

var b = B.new()
b.Foo()
```
I'm not sure what the best strategy here is. Thinking it's something not yet implemented rather than a bug. On an interface, class_itf2class needs to be examined. Some thoughts
- populate the interface's class_T directly with the information after compilation
- use a macro to get class_members_tv, it would handle seeing an interface

Examining itf2class might require knowing what extending class we're dealing with.

I'm going to try a `class C implements A` and see what itf2class looks like and I'll look for an example of class_itf2class used with methods.

Maybe you've got familiarity with this stuff, before I spend too much time looking around. (Read the following and the previous will make more sense...)

It crashes in `copy_tv` from `exec_instructions()` doing `case ISN_LOAD_CLASSMEMBER:`with

```
(gdb) p iptr->isn_arg.classmember.cm_class.class_members_tv 
$7 = (typval_T *) 0x0

and

p *iptr->isn_arg.classmember.cm_class
$10 = {
  class_name = 0x555555d31b00 "A",
  class_flags = 1,
  ...
```

In struct class_S there's
```
    itf2class_T	*class_itf2class;	// member index lookup tables
```
and
```
(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class 
$12 = {i2c_next = 0x555555d39a70, i2c_class = 0x555555d39960, i2c_is_method = 1}
(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class.i2c_next 
$13 = {i2c_next = 0x0, i2c_class = 0x555555d39960, i2c_is_method = 0}

```
found the class
```
(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class.i2c_next.i2c_class 
$14 = {class_name = 0x555555d3b330 "B", class_flags = 0,
```
and the value
```
(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class.i2c_next.i2c_class.class_members_tv 
$15 = {v_type = VAR_NUMBER, v_lock = 0 '\000', vval = {v_number = 40,
```

**Reply by errael** [reply]
@yegappan Doh! The first thing to look at is how `object` members are handled as opposed to `class` members.

**Reply by mg979** [reply]
@errael you could start with the simpler (and more important) stuff, like the BUGs here (accessing static class members), since they are probably related, and we didn't agree on interface specification.
```
vim9script
class A
    this.ro_obj_var = 10
    public this.rw_obj_var = 20
    this._priv_obj_var = 30

    static ro_class_var = 40
    public static rw_class_var = 50
    static _priv_class_var = 60
endclass

def T()
    var a = A.new()
    echo a.ro_obj_var		# Works
    a.ro_obj_var = 0		# Fails with E1335 (expected)
    echo a.rw_obj_var		# Works
    a.rw_obj_var = 0		# Works
    echo a._priv_obj_var	# Fails with E1333 (expected)
    a._priv_obj_var = 0		# Fails with E1333 (expected)
    
    echo A.ro_class_var		# Works
    A.ro_class_var = 0		# Fails with E1089 (BUG)
    echo A.rw_class_var		# Works
    A.rw_class_var = 0		# Fails with E1089 (BUG)
    echo A._priv_class_var	# Works (BUG)
    A._priv_class_var = 0	# Fails with E1089 (BUG)
enddef
T()
```

**Reply by errael** [reply]
I think I see the issue, at least partly.

An interface's members have storage (typeval_T) allocated with the object;
for static members, there would presumable only be one copy of the data
and the typeval_T would be allocated with the the interface not the object.

Currently there is storage for the statics allocated with the object

I set up
```
    B implements A
        static ro_class_var1 = 40
        static ro_class_var2 = 50
    C implements A
        static ro_class_var2 = 66
        static ro_class_var1 = 77
```
Note the reversed the order of the static declarations in `B` and `C`
It crashed after
```
var b = B.new()
b.Foo()
```
as expected. Note in the gdb output that the first typval_T in "B" is 40 which
is var1 and in "C" the first is "66" which is var2.

So, as suspected, each class has it's own storage of the statics and in their own order.

Not sure how much it would take to fix this, but one approach would be to
forbid/error static members in an interface until there's a fix. If a
programmer want to share data between objects, there are other simple ways
like a singleton class with the shared data.

Are you OK with giving an error for static in interface for now?
Where are interfaces declared?

```
(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class.i2c_next.i2c_next.i2c_next.i2c_class
$27 = {class_name = 0x555555d30e30 "B"...
(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class.i2c_next.i2c_next.i2c_next.i2c_class.class_members_tv
$28 = {v_type = VAR_NUMBER, v_lock = 0 '\000', vval = {v_number = 40,

(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class.i2c_next.i2c_class
$29 = {class_name = 0x555555d316e0 "C",...
(gdb) p *iptr->isn_arg.classmember.cm_class.class_itf2class.i2c_next.i2c_class.class_members_tv 
$30 = {v_type = VAR_NUMBER, v_lock = 0 '\000', vval = {v_number = 66
```
Here's the full program I used
```
vim9script

interface A
    static ro_class_var1: number
    static ro_class_var2: number
endinterface

class B implements A
    static ro_class_var1 = 40
    static ro_class_var2 = 50

    def Foo()
	var xxx = A.ro_class_var1
	var yyy = A.ro_class_var2
    enddef
endclass

class C implements A
    static ro_class_var2 = 66
    static ro_class_var1 = 77

    def Foo()
	var xxx = A.ro_class_var1
	var yyy = A.ro_class_var2
    enddef
endclass

#disassemble B.Foo
var b = B.new()
b.Foo()
```

**Reply by errael** [reply]
@yegappan I posted what I think is a complete analysis of the crash, with a suggestion on next step. Forgot to @ you (I get hit and miss delivery)

**Reply by yegappan** [reply]
Thanks for the detailed analysis.   Instead of preventing the usage of static members in interfaces, we should fix this crash.  Let me spend some time going through your analysis and the code.

**Reply by errael** [reply]
>  Instead of preventing the usage of static members in interfaces, we should fix this crash.

I agree. But since it looks like a development task, not just a bug fix, preventing it is a fallback while the bugs are rooted out. (hope I'm wrong).

**Reply by yegappan** [reply]
After going through your analysis and the code, it now makes sense to give an error when an interface member variable is referenced using the interface name.  A member variable in an interface cannot be initialized.  So it doesn't makes sense to allow read or write to this variable.
Can you create a PR for this?

**Reply by errael** [reply]
> So it doesn't makes sense to allow read or write to this variable.
> Can you create a PR for this?

(I'm still not entirely comfortable with how statics works, but I understand now)

What sounds like a good error message? How about
```
Cannot directly access static through interface.
```
or something like
```
Only access a static through a class, never an interface
````

Just thought of something else, and I'm still confused. What about
```
vim9script

interface I
    public static s_var: number
endinterface
class C implements I
    public static s_var = 3
endclass

def F(i: I)
    var x = i.s_var
enddef
var c = C.new()
F(c)
```
If that doesn't work, then interfaces don't work.

**Reply by errael** [reply]
BTW, the compilation error is
```
E1326: Member not found on object "I": s_var
```
for
```
def F(i: I)
    var x = i.s_var
enddef
```

**Reply by errael** [reply]
I may be getting back to the "disallow statics in interfaces in this release" point of view.

**Comment by errael** [comment]
I was asked to look at the crash; I'll finish what I'm doing, trying to track down the specifics of the crash, but maybe not a solution. He had something to look at after the crash; so I'll probably do that. As far as I'm concerned there should be a central figure coordinating, to prevent interference.

I do wonder why you felt it necessary to post that in the middle of the thread about the crash, rather than starting a separate thread. It's rude.

**Reply by mg979** [reply]
Sorry I didn't mean to be rude, just thought that starting from the hardest problem to solve maybe wasn't ideal. It was just a suggestion.

**Reply by errael** [reply]
> just a suggestion.

Which is OK with me. But a separate thread seems more appropriate.

**Comment by errael** [comment]
> A member variable in an interface cannot be initialized. 
> So it doesn't makes sense to allow read or write to this variable.
> Can you create a PR for this?

@yegappan #13007 

But until the issue of `def F(i: I)` is resolved, it might not make sense to apply this. Except as a last ditch effort to avoid a crash.

**Comment by yegappan** [comment]
The member access control issues discussed above are all addressed in the latest version of Vim.

**Comment by errael** [comment]
> A member variable in an interface cannot be initialized.
> So it doesn't makes sense to allow read or write to this variable.
> Can you create a PR for this?

#13007 prevents direct access to interface statics. It also allows a `:def` outside of class to read statics.
